chore(docs): v2 execution API spec (#1995)
astriaorg/astria: specs/execution-api.md

Reviewer: I think we should do this as a new file instead of replacing the old file.

Author: The old spec was already somewhat outdated, should we update it as well and put in a `deprecated` directory or something similar?

Author: I've gone ahead and put the old tx flow diagram and execution API spec in `deprecated` directories in b0bd402

ExecutionSessionParameters.celestia_search_height_max_look_ahead`.

## Rollup Implementation Details

### CreateExecutionSession

`CreateExecutionSession` returns an `ExecutionSession`, defining all necessary information
for the client to begin driving execution. This includes `ExecutionSessionParameters`,
containing the necessary information for network connection and mapping rollup blocks
to Astria Sequencer (soft) and Celestia (firm) blocks. Also returned is the current
`CommitmentState` of the server and a `session_id` to be supplied in RPC requests
for the current session.

This RPC should **always** succeed. The API is agnostic as
to how this information is defined in a rollup's genesis, and is only used by the
Conductor as configuration at the beginning of each execution session.

### ExecuteBlock

`ExecuteBlock` executes a set of given transactions on top of the chain
indicated by `prev_block_hash`. The following should be respected:

- `prev_block_hash` MUST match hash of the most recently committed block, whether
  `soft` or `firm`. RPC should return `FAILED_PRECONDITION` status otherwise.
- If block headers have timestamps, the created block MUST have matching timestamp

> - **NOTE:** The `CommitmentState` is NOT modified by the execution of the block.

Reviewer: Need to update to add information on the `sequencer_block_hash` which has since been added here

the Sequencer or the Celestia height). If the block cannot be found, return a `NOT_FOUND`
error.

### UpdateCommitmentState

`UpdateCommitmentState` replaces the current `CommitmentState` on the rollup.

- No `UpdateCommitmentStateRequest` can ever decrease in either `soft` or `firm`
  block number.
- `soft` blocks **must** increase in block number each time this is called.
- `firm` blocks **must** either increase in block number ***or*** match the current
  commitment state block.
- Block numbers in state **must** be such that  `soft` >= `firm`.
- If any of these conditions fail, a `FAILED_PRECONDITION` status should be returned.

## Sequence Diagram

The sequence diagram below shows the API used within the full context of Astria
stack, demonstrating what happens between when a user submits a transaction and
when they see it executed, as well as the process of soft and firm commitments.
Note that this diagram presumes the Conductor is running in `SoftAndFirm` mode,
driving execution when it receives a soft block. If it were running in `FirmOnly`
mode, `ExecuteBlock` would not be called until the block is received from the
Data Availability layer.

> ![image](assets/execution_api_sequence.png)

Reviewer: We should probably update the sequence diagram.

Author: How would you want to update it? As is, the only slight inaccuracy that I can see is that rollup block is returned by `ExecuteBlock` instead of `ExecutedBlockMetadata`. Should we add the `CreateExecutionSession` call in the "setup" portion?

Reviewer: Yeah we should add the execution session and make sure the naming is all correct.

Author: Updated in b0bd402

# Execution API (v2) Specification

## Overview

The Execution API is the interface `Conductor` uses to drive deterministic derivation
of a rollup chain from Sequencer and Celestia blocks. Inspired by other APIs, such
as the Engine API and ABCI Consensus API, it is a chain-agnostic mechanism intended
to be very simple to implement. It is a gRPC API which any state machine can implement
and use Conductor with to drive their block creation to integrate with the Astria
Sequencer.

## Basic Design Principles

The Execution API is a resource-based API with three resources: `ExecutedBlockMetadata`,
`CommitmentState`, and `ExecutionSession`. The API is designed to follow basic
principles outlined by aip.dev as best practices for resource based APIs. gRPC
has been chosen for the API due to the wide availability of language implementations
which make it easy to generate client libraries and server implementations.

## Conductor Usage

### Execution Sessions

> Client driven execution of blocks occurs in execution "sessions", which represent

Author: Done in 3f928ef

a bound of block heights/numbers to be executed. Every session contains a lower

> bound (start block) and may also contain an upper bound (end block). Calls to RPCs

Author: Good catch on the RPCs, done in 3f928ef.

within the current session can be verified by the server-side application via the
`session_id` in `GetExecutedBlockMetadataRequest` and `ExecuteBlockRequest`.

If an upper bound to the current execution session is specified in the `ExecutionSessionParameters`,

> the execution client will end the session after executing the end block and start
> a new session via `CreateExecutionSession`. This allows for changes to client configuration
> (such as chain ID, height mapping, etc.) between sessions.

Reviewer 2: We decided against providing a mechanism to outright end a session so I would rephrase this slightly (also added a example useage for the new session):
```suggestion
the execution client stop executing past the end block number and request
a new session via `CreateExecutionSession`. This allows providing the client with new
session parameters (such as chain ID, height bounds, etc.) and can be used to perform
rolling hard forks.
```

Author: Changed `client` to `Conductor` to follow precedent you mentioned above, made these changes in 3f928ef.

### Startup

Upon startup, conductor starts its first execution session by calling `CreateExecutionSession`.
This returns an `ExecutionSession`, containing the necessary information for the
client to drive execution for the duration of the session.

### Execution & Commitments

From the perspective of the Conductor:

- `Soft` commitments have been fully committed by the Sequencer consensus.
- A `Firm` commitment indicates that the block has been written and propagated

>   across the DA network.

Reviewer 2: ```suggestion
across Celestia.
```

Author: Done in 3f928ef

When configuring Conductor, the threshold at which blocks are executed on the rollup
can be set via the `execution_commitment_level` in the config file. `ExecuteBlock`
is called to create a new rollup block when the `execution_commitment_level` has
been reached for a given block. Upon receipt of the executed block, Conductor calls
`UpdateCommitmentState` to update the commitment at the level of the
`execution_commitment_level` and any level above it.

`execution_commitment_level` options and changes to execution:

> - `SoftOnly`

Reviewer 2: Sort these in a different order because it's easier to demonstrate the differences:

1. `SoftOnly`
2. `FirmOnly`
3. `SoftAndFirm`

Author: Done in 3f928ef.

  - upon receiving a new sequencer block N from the Sequencer:
    - `ExecuteBlock` will be called with data from the sequencer block N, then
    - `UpdateCommitmentState` will be called to update the `soft` block to N
- `SoftAndFirm`
  - upon receiving a new sequencer block N from the Sequencer:
    - `ExecuteBlock` will be called with data from the sequencer block N, then
    - `UpdateCommitmentState` will be called to update the `soft` block to N

>   - upon reading new blocks from DA containing all of blocks K->N, where K is

Reviewer 2: We should talk about Celestia here. We don't use `DA` anywhere else, and we also don't have any alternatives.

```suggestion
- upon reading new blocks from Celestia containing all of blocks K->N, where K is
```

    some arbitrary ancestor of N
    - `UpdateCommitmentState` will be called to update the `firm` block to N
- `FirmOnly`
  - conductor does not need to listen for new blocks from Sequencer

>   - upon reading new blocks from DA containing all blocks K->N

Reviewer 2: ```suggestion
  - upon reading new blocks from Celestia containing all blocks K->N
```

I feel like we should probably get rid of the notion of `blocks K->N` here because we don't really go into sufficient detail here, which makes this very confusing.

Let's just leave it at the following (and don't go into the `For each` loop below):

```suggestion
  - upon reading a new Sequencer block N from Celestia:
```

Author: Done in 3f928ef. Would you also want a similar change to the `SoftAndFirm` section, or is that one clear enough?

Reviewer 2: Yes, it would be good to also fix it there because it's just as confusing. :D (I think I wrote this stuff back in the day....)

Author: Done in 7c9ee1c

    - For each block M from K->N:
      - `ExecuteBlock` will be called with data from the sequencer block M
      - `UpdateCommitmentState` will be called to update `firm` and `soft` blocks
        to M

Note: For our EVM rollup, we map the `CommitmentState` to the `ForkchoiceRule`:

- `Soft` Commitment -> `HEAD` Forkchoice && `SAFE` Forkchoice
- `Firm` Commitment -> `FINAL` Forkchoice

> ### Celestia Search Height

Reviewer 2: This might be an appropriate place to go into the details of how Astria Sequencer block data for a given rollup is read from Celestia.

IMO call this section:

```
### Reading Rollup-specific Astria block data from Celestia
```

(Or simpler "Reading Astria blocks from Celestia").

current implementation of Conductor, this is set to the Celestia height at which
the most recent firm block was found.

There are, however, many factors that can result in the same Sequencer blocks sharing
or skipping Celestia heights. As such, the Celestia heights correlating to Sequencer
commitments may not increase linearly, and a range of heights must be searched.
Conductor begins fetching at `CommitmentState.lowest_celestia_search_height`, continuing
searching for firm commitments until it reaches `lowest_celestia_search_height +
ExecutionSessionParameters.celestia_search_height_max_look_ahead`.

## Rollup Implementation Details

### CreateExecutionSession

`CreateExecutionSession` returns an `ExecutionSession`, defining all necessary information
for the client to begin driving execution. This includes `ExecutionSessionParameters`,
containing the necessary information for network connection and mapping rollup blocks
to Astria Sequencer (soft) and Celestia (firm) blocks. Also returned is the current
`CommitmentState` of the server and a `session_id` to be supplied in RPC requests
for the current session.

> This RPC should **always** succeed. The API is agnostic as

Reviewer 3: This is not true even for our own implementation. `CreateExecutionSession` does not always succeed. We fail with `FAILED_PRECONDITION` or `INTERNAL` depending on the specific issue.

Author: Removed in 3f928ef

to how this information is defined in a rollup's genesis, and is only used by the
Conductor as configuration at the beginning of each execution session.

> 

Reviewer 3: would add that calls to `CreateExecutionSession` MUST invalidate any previous sessions.

Author: Done in 3f928ef

### ExecuteBlock

`ExecuteBlock` executes a set of given transactions on top of the chain
indicated by `parent_hash`. The following should be respected:

- `parent_hash` MUST match hash of the most recently committed block, whether

>   `soft` or `firm`. RPC should return `FAILED_PRECONDITION` status otherwise.

Reviewer 3: would change to "RPC MUST return `FAILED_PRECONDITION` ..."

Author: Done in 3f928ef

- If block headers have timestamps, the created block MUST have matching timestamp
- **NOTE:** The `CommitmentState` is NOT modified by the execution of the block.
- It is up to the execution node if it includes the `sequencer_block_hash`
  provided as a part of the executed block metadata. If utilized, the server **MUST**
  throw an `INVALID_ARGUMENT` error if the `sequencer_block_hash` is not included
  in the request.

> 

Author: Done in 3f928ef

### GetExecutedBlockMetadata

`GetExecutedBlockMetadata` returns information about a block given its `ExecutedBlockIdentifier`,
consisting of either a `hash` or `number` (as determined by the server, *not* by
the Sequencer or the Celestia height). If the block cannot be found, return a `NOT_FOUND`
error.

### UpdateCommitmentState

`UpdateCommitmentState` replaces the current `CommitmentState` on the rollup.

- No `UpdateCommitmentStateRequest` can ever decrease in either `soft` or `firm`
  block number.
- `soft` blocks **must** increase in block number each time this is called.
- `firm` blocks **must** either increase in block number ***or*** match the current
  commitment state block.
- Block numbers in state **must** be such that  `soft` >= `firm`.

> - If any of these conditions fail, a `FAILED_PRECONDITION` status should be returned.

Reviewer 3: ... a `FAILED_PRECONDITION` status MUST be returned

Author: Done in 3f928ef

> 

Reviewer 3: would add the following (not necessarily this exact verbiage):
- MUST return `PERMISSION_DENIED` if called without a valid session
- MUST return `OUT_OF_RANGE` if the soft block commitment is out of bounds for the session
- MUST return `OUT_OF_RANGE` if the firm block commitment does not match the current commitment state block and is out of bounds for the session

Author: Done in 3f928ef

- A `Firm` commitment indicates that the block has been written and propagated
  across Celestia.

When configuring Conductor, the threshold at which blocks are executed on the rollup
can be set via the `execution_commitment_level` in the config file. `ExecuteBlock`
is called to create a new rollup block when the `execution_commitment_level` has
been reached for a given block. Upon receipt of the executed block, Conductor calls
`UpdateCommitmentState` to update the commitment at the level of the
`execution_commitment_level` and any level above it.

`execution_commitment_level` options and changes to execution:

- `SoftOnly`
  - upon receiving a new sequencer block N from the Sequencer:
    - `ExecuteBlock` will be called with data from the Sequencer block N, then
    - `UpdateCommitmentState` will be called to update the `soft` block to N
- `FirmOnly`
  - conductor does not need to listen for new blocks from Sequencer
  - upon reading a new Sequencer block N from Celestia:
    - `ExecuteBlock` will be called with data from the Sequencer block N
    - `UpdateCommitmentState` will be called to update `firm` and `soft` blocks
      to N

> - `SoftAndFirm`

Reviewer 2: Add a point below here that regular operation of Conductor assumes that Soft blocks/blocks from Sequencer are read before Firm blocks/blocks from Celestia - but that conductor can deal with the reverse scenario.

Author: Sorry for the commit spam, this comment only appeared when I reloaded the page. Done here: 1df5eb7.

>   - upon receiving a new Sequencer block N from the Sequencer:

Reviewer 2: Maybe add here:

```
    - if Sequencer block `N` has not yet been read from Celestia (this is considered normal operation):
      - <execute block && updatecommitmentstate>
    - if Sequencer block `N` has already been read from Celestia:
      - <skip block>
```

    - `ExecuteBlock` will be called with data from the Sequencer block N, then
    - `UpdateCommitmentState` will be called to update the `soft` block to N

>   - upon reading Sequencer block N from Celestia:

Reviewer 2: And similar to my comment above:

```
    - if Sequencer block N has already been read from the Sequencer (this is expected to be regular operation):
      - <updatecommitmentstate>
    - if Sequencer block N has not yet been read from Sequencer:
      - <execute && updatecommitmentstate firm>
```

Author: Done in af66f3a.
